lib/SILOptimizer/Utils/Devirtualize.cpp

```diff
@@ -874,7 +874,9 @@ getWitnessMethodSubstitutions(
       witnessThunkSig);
 }
 
-static SubstitutionMap
+// This function is made public so it can be used by the constexpr propagation
```

Comment:
Instead, why not make a public function which devirtualizes an apply of a witness_method, and then the constant evaluator only has to worry about calls of static function references?

Reply:
Mandatory inlining has something similar as well, so at this point it looks like this is adding a fourth copy of a very similar piece of code.

Reply:
I set out to do this and discovered an obstacle: I can't actually mutate the SIL to have static function references (like the other APIs in Devirtualize.h do) because the function containing the apply can still be totally generic. The interpreter knows all the concrete types because it arrived at the apply by following a specific path.

I could create a Devirtualize.h public function that takes an ApplyInst and the current substitutions and returns a devirtualized ApplyInst without replacing the existing one. That might be confusingly different from the other Devirtualize.h APIs, and it also wastefully allocates an ApplyInst that is used for a very temporary purpose.

Reply:
What's your suggestion?

Reply:
@ravikandhadai @marcrasi I think refactoring the devirtualizer to return a `SILFunction *` together with a SubstitutionMap might be the way to go. You're right that allocating a new ApplyInst might be wasteful if you're not actually mutating the SIL.